pkg/tournament/classic_division.go

```diff
@@ -481,171 +585,34 @@ func newPlayerIndexMap(players []string) map[string]int {
-		l := len(standings)
-		for i := 0; i < l-1; i += 2 {
-			playerOne := standings[i].Player
-			playerTwo := standings[i+1].Player
-			var newPairing *entity.Pairing
-			// If we are past the first round in an elimination tournament,
-			// the bottom half of the standings have been eliminated.
-			// Each successive round eliminates half as many players,
-			// hence the l / twoPower(l) determines which players are eliminated.
-			if pairingMethod == entity.Elimination && round > 0 && i >= l/twoPower(round) {
-				newPairing = newEliminatedPairing(playerOne, playerTwo)
-			} else {
-				newPairing = newClassicPairing(t, playerOne, playerTwo, round)
-			}
-			roundPairings[t.PlayerIndexMap[playerOne]].Pairing = newPairing
-			roundPairings[t.PlayerIndexMap[playerTwo]].Pairing = newPairing
-		}
-	} else if pairingMethod == entity.Random {
-		playerIndexes := []int{}
-		for _, v := range t.PlayerIndexMap {
-			playerIndexes = append(playerIndexes, v)
-		}
-		// rand.Seed(time.Now().UnixNano())
-		rand.Shuffle(len(playerIndexes),
-			func(i, j int) {
-				playerIndexes[i], playerIndexes[j] = playerIndexes[j], playerIndexes[i]
-			})
-		for i := 0; i < len(playerIndexes)-1; i += 2 {
-			playerOne := t.Players[playerIndexes[i]]
-			playerTwo := t.Players[playerIndexes[i+1]]
-			newPairing := newClassicPairing(t, playerOne, playerTwo, round)
-			roundPairings[playerIndexes[i]].Pairing = newPairing
-			roundPairings[playerIndexes[i+1]].Pairing = newPairing
-		}
-	} else if pairingMethod == entity.RoundRobin {
-		roundRobinPlayers := t.Players
-		// The empty string represents the bye
-		if len(roundRobinPlayers)%2 == 1 {
-			roundRobinPlayers = append(roundRobinPlayers, "")
-		}
-
-		roundRobinPairings := getRoundRobinPairings(roundRobinPlayers, round)
-		for i := 0; i < len(roundRobinPairings)-1; i += 2 {
-			playerOne := roundRobinPairings[i]
-			playerTwo := roundRobinPairings[i+1]
-
-			if playerOne == "" && playerTwo == "" {
-				return errors.New(fmt.Sprintf("Two byes playing each other in round %d\n", round))
-			}
-
-			// The blank string represents a bye in the
-			// getRoundRobinPairings algorithm, but in a Tournament
-			// is represented by a player paired with themselves,
-			// so convert it here.
-
-			if playerOne == "" {
-				playerOne = playerTwo
-			}
-
-			if playerTwo == "" {
-				playerTwo = playerOne
-			}
-
-			newPairing := newClassicPairing(t, playerOne, playerTwo, round)
-			roundPairings[t.PlayerIndexMap[playerOne]].Pairing = newPairing
-			roundPairings[t.PlayerIndexMap[playerTwo]].Pairing = newPairing
-		}
-	}
-	// Give all unpaired players a bye
-	// realtime.TournamentGameResult_BYEs are always designated as a player
-	// paired with themselves
-	if pairingMethod != entity.Manual {
-		for i := 0; i < len(roundPairings); i++ {
-			pri := roundPairings[i]
-			if pri.Pairing == nil {
-				pri.Pairing = newClassicPairing(t, t.Players[i], t.Players[i], round)
+	repeats := make(map[string]int)
+	for i := 0; i <= round; i++ {
```

Comment:
The only call to getRepeats uses round - 1, so round is always guaranteed to be round < len(t.Matrix) - 1. I decided to decrement the int in the call instead of using '<' in getRepeats. I thought it made more sense for the call to getRepeats to be something like "Get all repeats up to and **including** this round". Also, we might want to get the repeats for the whole tournament at some point (even though getRepeats(t, len(t.Matrix)-1) would be useless for pairings since the tournament would be over), and having and '<=' made more sense to me rather than passing in getRepeats(t, len(t.Matrix)) (at least for me). Let me know what you think.

Reply:
The negative round check is redundant though, I will remove that.

Reply:
i misread this code 😅

[exclusive upper bound](https://tour.golang.org/moretypes/7) is more common, so i try to stick to that since i always forget a +1 or -1 somewhere. without context, i'd read "getRepeats(t, n)" as "get repeats within first n rounds" rather than "get repeats from round 0 to round n (zero-based)".

but if "inclusive" is intended here, then it looks like this code is fine, except when `len(t.Matrix()) == 0 && round < 0`. can the `if round < 0 { round = 0 }` be removed?

Reply:
Yes, the zero can be (and has been) removed.